src/b3d/chisight/gen3d/image_kernel.py

```diff
@@ -310,3 +310,98 @@ def get_rgbd_vertex_kernel(self) -> PixelRGBDDistribution:
+            (
+                hyperparams["intrinsics"]["image_height"].unwrap(),
+                hyperparams["intrinsics"]["image_width"].unwrap(),
+                4,
+            )
+        )
+
+    def logpdf(
+        self, observed_rgbd: FloatArray, state: Mapping, hyperparams: Mapping
+    ) -> FloatArray:
+        intrinsics = hyperparams["intrinsics"]
+
+        transformed_points = state["pose"].apply(hyperparams["vertices"])
+
+        # Sort the vertices by depth.
+        sort_order = jnp.argsort(transformed_points[..., 2])
+        transformed_points_sorted = transformed_points[sort_order]
+
+        # Project the vertices to the image plane.
+        projected_coords = jnp.rint(
+            b3d.utils.xyz_to_pixel_coordinates(
+                transformed_points_sorted,
+                intrinsics["fx"],
+                intrinsics["fy"],
+                intrinsics["cx"],
+                intrinsics["cy"],
+            )
+            - 0.5
+        ).astype(jnp.int32)
+
+        # Compute the unique pixel coordinates and the indices of the first vertex that hit that pixel.
+        unique_pixel_coordinates, unique_indices = jnp.unique(
+            projected_coords,
+            axis=0,
+            return_index=True,
+            size=projected_coords.shape[0],
+            fill_value=-1,
+        )
+
+        # For each collided pixel, get the observed rgbd value.
+        observed_rgbd_per_point = observed_rgbd[
+            unique_pixel_coordinates[:, 0], unique_pixel_coordinates[:, 1]
+        ]
+
+        # For each collided pixel, get the color and depth of the vertex that hit that pixel.
+        latent_rgbd_per_point = jnp.concatenate(
+            [
+                state["colors"][sort_order][unique_indices],
+                transformed_points_sorted[unique_indices, 2, None],
+            ],
+            axis=-1,
+        )
+
+        # Score the collided pixels
+        scores = jax.vmap(
+            hyperparams["image_kernel"].get_rgbd_vertex_kernel().logpdf,
+            in_axes=(0, 0, None, None, 0, 0, None),
+        )(
+            observed_rgbd_per_point,
+            latent_rgbd_per_point,
+            state["color_scale"],
+            state["depth_scale"],
+            state["visibility_prob"],
+            state["depth_nonreturn_prob"],
+            hyperparams["intrinsics"],
+        )
+        is_valid_pixel = (unique_pixel_coordinates >= 0).all(axis=-1)
+        adjusted_scores = jnp.where(is_valid_pixel, scores, 0.0)
+
+        # Calcuate the number of pixels that don't have any vertices hitting them.
+        total_pixels = (
+            intrinsics["image_height"].unwrap() * intrinsics["image_width"].unwrap()
+        )
+        num_no_hypothesis_pixels = total_pixels - is_valid_pixel.sum()
+
+        # The pixels that have no vertices produce an observation uniformly at random.
+        color_score = jnp.log(1 / 1.0**3)
+        depth_score = jnp.log(1 / (intrinsics["far"] - intrinsics["near"]))
```

Comment:
@nishadgothoskar this looks great!
Would it be easy to set this up to accumulate a different depth score for (1) pixels with no vertices, that have a depth nonreturn (ie. depth==0), and (2) pixels with no vertices, that have a valid depth observation?

The score for pixels of type (1) should be `log(p_depth_nonreturn)` and the score for pixels of type (2) should be `log(1 - p_depth_nonreturn) + log(1 / (far - near))`.  Here, `p_depth_nonreturn` is a value we should set (a model hyperparam).  In the generative process, it is the probability that an unexplained pixel random pixel gets a depth nonreturn, rather than a uniformly random depth.

I'm not totally sure, but I think there is a chance that if we don't have this, inference will sometimes be too unwilling to choose poses in which there are more depth nonreturn pixels not explained by the latent object.